build-tools/Xamarin.Android.Tools.BootstrapTasks/Xamarin.Android.Tools.BootstrapTasks/CreateFrameworkList.cs

```diff
@@ -0,0 +1,86 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Xml.Linq;
+
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+namespace Xamarin.Android.Tools.BootstrapTasks
+{
+	public class CreateFrameworkList : Task
+	{
+		[Required]
+		public                  ITaskItem       FrameworkListFile           { get; set; }
+
+		[Required]
+		public                  ITaskItem       FrameworkDirectory          { get; set; }
+
+		[Required]
+		public                  ITaskItem[]     FrameworkFiles              { get; set; }
+
+		public                  ITaskItem[]     FrameworkFileOverrides      { get; set; }
+
+		[Required]
+		public                  string          Redist                      { get; set; }
+
+		[Required]
+		public                  string          Name                        { get; set; }
+
+		public override bool Execute()
+		{
+			var files = new List<ITaskItem>(FrameworkFiles);
+			files.Sort ((x, y) => {
+				var a = Path.GetFileNameWithoutExtension (x.ItemSpec);
+				var b = Path.GetFileNameWithoutExtension (y.ItemSpec);
+				return string.Compare (a, b, StringComparison.OrdinalIgnoreCase);
+			});
+			var contents = new XElement ("FileList",
+					new XAttribute ("Redist", Redist),
+					new XAttribute ("Name", Name),
+					files.Select (f => ToFileElement (f)));
+			contents.Save (FrameworkListFile.ItemSpec);
+			return true;
+		}
+
+		XElement ToFileElement (ITaskItem file)
+		{
+			var path = Path.Combine (FrameworkDirectory.ItemSpec, file.ItemSpec);
+			if (!File.Exists (path)) {
+				path = Path.Combine (FrameworkDirectory.ItemSpec, "Facades", file.ItemSpec);
+			}
+			using (var pe = new PEReader (File.OpenRead (path))) {
+				var reader          = pe.GetMetadataReader();
+				var assemblyDef     = reader.GetAssemblyDefinition();
+
+				var taskVersion     = Nullable (file.GetMetadata ("Version"));
+				var overrideVersion = Nullable (FrameworkFileOverrides.FirstOrDefault (o => o.ItemSpec == file.ItemSpec)?.GetMetadata ("Version"));
+				var assemblyVersion = assemblyDef.Version.ToString ().TrimStart ('v');
+				var version         = taskVersion ?? overrideVersion ?? assemblyVersion;
+
+				var assemblyName    = new AssemblyName();
+				assemblyName.SetPublicKey (reader.GetBlobBytes (assemblyDef.PublicKey));
```

Comment:
`System.Reflection.Metadata.AssemblyDefinition` has a `GetAssemblyName()` method that returns `System.Reflection.AssemblyName`, can we use that instead?

Reply:
lol?  When I attempt to use that, it fails horrifically:

```
The "CreateFrameworkList" task failed unexpectedly.
System.MissingMethodException: Method not found: System.Reflection.AssemblyName System.Reflection.Metadata.AssemblyDefinition.GetAssemblyName()
  at Xamarin.Android.Tools.BootstrapTasks.CreateFrameworkList.<Execute>b__24_1 (Microsoft.Build.Framework.ITaskItem f)
  at System.Linq.Enumerable+SelectListIterator`2[TSource,TResult].MoveNext ()
  at System.Xml.Linq.XContainer.AddContentSkipNotify (System.Object content)
  at System.Xml.Linq.XContainer.AddContentSkipNotify (System.Object content)
  at System.Xml.Linq.XElement..ctor (System.Xml.Linq.XName name, System.Object content)
  at System.Xml.Linq.XElement..ctor (System.Xml.Linq.XName name, System.Object[] content)
  at Xamarin.Android.Tools.BootstrapTasks.CreateFrameworkList.Execute ()
  at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute ()
```

This is because on macOS, `msbuild` itself uses an older version `System.Reflection.Emit`, which we had to work around in commit 0d9473c152edfa30a35a5f5162bb9662821c5e6a/PR #2634.

Reply:
@radical we've been hitting this a lot ^^

Is `System.Reflection.Metadata.dll` something that will ship in Mono at some point? I don't see it on Windows/2019:

![image](https://user-images.githubusercontent.com/840039/55582068-25e7ef00-56e4-11e9-9950-662fc38b60c2.png)

Reply:
@jonpryor I filed: https://github.com/mono/mono/issues/13861